Cloud Monitoring: Out-of-the-box dashboards

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Adds out-of-the-box dashboards for the Cloud Monitoring datasource

These dashboards have been derived from the respective Monitoring ones (also available in [CGP console](https://console.cloud.google.com/monitoring/dashboards)):

- [BigQuery Query Monitoring](https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/master/dashboards/bigdata/bigquery-monitoring.json)
- [Cloud SQL Monitoring](https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/master/dashboards/storage/cloudsql-monitoring.json)
- [GCE VM Instance Monitoring](https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/master/dashboards/compute/gce-vm-instance-monitoring.json)
- [GKE Cluster Monitoring](https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/master/dashboards/compute/gke-cluster-monitoring.json)
- [HTTP_S Load Balancer Monitoring](https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/master/dashboards/networking/https-loadbalancer-monitoring.json)

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

- [X] A complementary PR adding the new image for the documentation will follow up.

papagian committed:
Cloud moniroring: out-of-the-box dashboards

wbrowne commented on docs/sources/features/datasources/cloudmonitoring.md:
```diff
@@ -335,3 +335,22 @@ Only available in Grafana v7.1+.
+
+> Only available in Grafana v7.3+.
+
+The updated Cloud Monitoring data source ships with pre-configured dashboards for five of the most popular GCP services:
```
Should `data source` be `datasource` in this case? 🤔 Same for below on line 351

oddlittlebird replied:
Yes, per the [Word usage section of the Grafana documentation style guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md#setup-set-up).

marcusolsson replied:
From the [style guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md#data-source):

> Two words, not one

wbrowne commented on public/app/plugins/datasource/cloud-monitoring/plugin.json:
```diff
@@ -7,6 +7,17 @@
   "alerting": true,
   "annotations": true,
+  "includes": [
+    { "type": "dashboard", "name": "BigQuery Query Monitoring", "path": "dashboards/BigQuery Query Monitoring.json" },
```
I'm wondering if it's safer to have filenames without spaces?

oddlittlebird replied:
@marcusolsson, are there plugin-specific recommendations? If not, should there be?

marefr replied:
Without spaces and lowercase is probably the safest

papagian replied:
thank you all for your comments; I will change them to lowercase without spaces.
I'm thinking also that `/` separator can also be an issue in Windows but I don't know how to change it to work for both. This must be [where](https://github.com/grafana/grafana/blob/master/pkg/plugins/dashboards.go#L99) it is consumed. Any ideas?

papagian replied:
I have tested importing dashboards in windows (from the [cloudwatch datasource](https://github.com/grafana/grafana/blob/b892aa2eb73f64f49f217f6f7c74f37cd9e73959/public/app/plugins/datasource/cloudwatch/plugin.json#L11)) and it works so the `/` separator is not a problem (as I expected).

oddlittlebird commented on docs/sources/features/datasources/cloudmonitoring.md:
```diff
@@ -335,3 +335,22 @@ Only available in Grafana v7.1+.
+
+## Out-of-the-box dashboards
+
+> Only available in Grafana v7.3+.
```
Please do not merge this until the 7.3 beta. @daniellee , FYI.

marefr replied:
Why? Have we changed this approach? Not optimal having PR's ready, but not merged for some time and then they're going to be merged for the beta. We want to merge as fast possible to make sure it works as expected in master. Guess the approach would be to open separate PR's for documentation then?

oddlittlebird replied:
Literally having a meeting today to discuss it, yes. And I'm just talking about docs, not the feature.

Which is not ideal, I know. I would like the docs to stay in lockstep with the code, just as this PR has it, but right now we are trying to figure out how to keep the future-state stuff out of the current docs as much as possible (currently 7.2 docs) without driving us all insane.

torkelo replied:
Until we have solved https://github.com/grafana/grafana/issues/27384 we are stuck with version notices (which can be good even post that)

marefr replied:
@papagian which means we can merge this before 7.3 beta leaving the version notes intact.

oddlittlebird commented on docs/sources/features/datasources/cloudmonitoring.md:
```diff
@@ -335,3 +335,22 @@ Only available in Grafana v7.1+.
+
+To import the pre-configured dashboards, go to the configuration page of a Cloud monitoring data source and click on the `Dashboards` tab. Click `Import` for the dashboard you would like to use.
+The datasource of the newly created dashboard panels will be the one selected above.
+The dashboards have a template variable which is populated with the projects accessible by the configured service account every time the dashboard is loaded. After the dashboard is loaded, you can select the project you prefer from the drop down list.
```
```suggestion
The dashboards have a template variable which is populated with the projects accessible by the configured service account every time the dashboard is loaded. After the dashboard is loaded, you can select the project you prefer from the drop-down list.
```

papagian committed:
Remove redundant information related to dev account

papagian committed:
Update docs/sources/features/datasources/cloudmonitoring.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

papagian committed:
Rename JSON files

achatterjee-grafana commented on docs/sources/features/datasources/cloudmonitoring.md:
```diff
@@ -335,3 +335,22 @@ Only available in Grafana v7.1+.
+
+The updated Cloud Monitoring data source ships with pre-configured dashboards for five of the most popular GCP services:
+
+- BigQuery Query Monitoring,
```
Remove comma after bullets. No period after the last bullet.

achatterjee-grafana replied:
Also, the topic should have lazy numbering for all procedure sections. Need correction.

papagian replied:
what do you mean by "lazy numbering". Can you be more specific or point me to an example? thanks!

achatterjee-grafana replied:
All numbering should be 1, and not 1,2,3, ... they will render correctly in your preview.

Here is an explanation:

Use lazy numbering for long lists
Markdown is smart enough to let the resulting HTML render your numbered lists correctly. For longer lists that may change, especially long nested lists, use “lazy” numbering:

1.  Foo.
1.  Bar.
    1.  Foofoo.
    1.  Barbar.
1.  Baz.

papagian committed:
Fix docs

papagian committed:
Fixe docs

marefr commented on public/app/plugins/datasource/cloud-monitoring/dashboards/loadBalancer.json:
```diff
@@ -0,0 +1,1132 @@
+  },
+  "timepicker": {},
+  "timezone": "",
+  "title": "HTTP/S Load Balancer Monitoring",
```
Compared to the other title which are very GCP centrix, this one is more general and may clash with users existing dashboards. Not sure how big a problem that is though.

papagian replied:
I just used the original [title](https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/master/dashboards/networking/https-loadbalancer-monitoring.json#L2).
What about `HTTP/S GCP Load Balancer Monitoring`?

marefr reviewed (approved):
LGTM, but left a comment/questions.

Have a general question as well. Has there been any discussions about using a data source variable to support multiple data source instances? Now you can only import the dashboard and use for that specific data source you're importing from. Besides selecting a project, being able to select which Cloud Monitoring data source in a drop down before project could be useful, I think that's how CloudWatch dashboards have been implemented.

papagian commented:
> LGTM, but left a comment/questions.
> 
> Have a general question as well. Has there been any discussions about using a data source variable to support multiple data source instances? Now you can only import the dashboard and use for that specific data source you're importing from. Besides selecting a project, being able to select which Cloud Monitoring data source in a drop down before project could be useful, I think that's how CloudWatch dashboards have been implemented.

Initially I had such a datasource template variable for selecting datasource as well but I change it
because in order to import such a dashboard you need to select first a cloud monitoring datasource
(and then navigating to the Dashboards tab)
it made more sense to me to associate the dashboard to the specific datasource.
However, I'm open to change it this it we decide that is more useful.

marefr commented:
> Initially I had such a datasource template variable for selecting datasource as well but I change it
because in order to import such a dashboard you need to select first a cloud monitoring datasource
(and then navigating to the Dashboards tab)

Not sure I follow. Given you have one or several cloud monitoring data sources configured you can go into one of these, dashboards tab and import the dashboards and it will work for all your cloud monitoring data sources.

> it made more sense to me to associate the dashboard to the specific datasource.
> However, I'm open to change it we decide that is more useful.

I guess the limitation with this approach is that you can only use the dashboards with one data source at a time. Importing a dashboard for data source A it will work for data source A. Importing it again from data source B it will stop working for data source A and work for data source B.

papagian committed:
Parameterized dashboard by CM datasource

papagian committed:
Merge remote-tracking branch 'origin/master' into cm-ootb-dashboards

papagian commented:
> > Initially I had such a datasource template variable for selecting datasource as well but I change it
> > because in order to import such a dashboard you need to select first a cloud monitoring datasource
> > (and then navigating to the Dashboards tab)
> 
> Not sure I follow. Given you have one or several cloud monitoring data sources configured you can go into one of these, dashboards tab and import the dashboards and it will work for all your cloud monitoring data sources.
> 
> > it made more sense to me to associate the dashboard to the specific datasource.
> > However, I'm open to change it we decide that is more useful.
> 
> I guess the limitation with this approach is that you can only use the dashboards with one data source at a time. Importing a dashboard for data source A it will work for data source A. Importing it again from data source B it will stop working for data source A and work for data source B.

@marefr
I updated the dashboards and I've added the datasource variable to support multiple data source instances
do you want to have a fresh look?

marefr reviewed (approved):
LGTM

papagian commented:
@oddlittlebird
doc modifications look good to you now?
If yes, can you approve them? thanks!

papagian committed:
Fix metric kind in CloudSQL dashboard

papagian committed:
Merge remote-tracking branch 'origin/master' into cm-ootb-dashboards

papagian committed:
Set alias to instance name for GCE dashboard panels

achatterjee-grafana reviewed (approved):
Approving, assuming the changes Diana(@oddlittlebird) and Marcus asked for, were made. Thank you!
